### PR TITLE
[BEAM-12164]  Support querying against Postgres for the SpannerIO change streams connector

### PR DESCRIPTION
**Pull request by nancyxu123**
Support Postgres databases for the SpannerIO change stream connector. The connector should be able to query when (1) both the metadata database and the change streams database are GSQL (2) when both the metadata database and the change streams database are Postgres (3) when the metadata database is in Postgres and the change streams database is in GSQL and (4) when the metadata database is in GSQL and the change streams database is in Postgres

This PR first does an initial lookup to determine whether or not the change streams database and the metadata database are Postgres in SpannerIO. Then, it makes different queries to the metadata table and to the change stream depending on whether or not the database is postgres. It also creates the metadata table differently depending on whether the metadata database is Postgres. Finally, it maps received records into ChangeStreamRecords differently depending on whether or not the change stream is in a Postgres database or not.

### CODE REVIEW

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java:1632]: Maybe log the database id + dialect here and below?

```
DatabaseId changeStreamDatabaseId = DatabaseId.of(changeStreamConfig.getProjectId(), changeStreamConfig.getInstanceId(), changeStreamConfig.getDatabaseId);
Dialect changeStreamDatabaseDialect = getDialect(changeStreamSpannerConfig) ? 
LOG.info("The Spanner database " + changeStreamDatabaseId + " has dialect " + changeStreamDatabaseDialect);
```

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java:1709]: You can use `databaseClient.getDialect()` and remove this method

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java:1647]: I think you should give it a [Dialect](https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/main/java/com/google/cloud/spanner/Dialect.java), which makes the class more extensible.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java:1658]: I think you should it a [Dialect](https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/main/java/com/google/cloud/spanner/Dialect.java), which makes the class more extensible.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java:1659]: Same as above

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/NameGenerator.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/NameGenerator.java:47]: Since GoogleSQL char limit is lesser than PG, could we always use that instead (to avoid passing in a dialect here?). If you feel strongly that we should use the max for PG, could you give this method a Dialect instead of a boolean?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/dao/ChangeStreamDao.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/dao/ChangeStreamDao.java:55]: Could we use a Dialect instead of a boolean?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/dao/DaoFactory.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/dao/DaoFactory.java:69]: Could we use a `Dialect` instead of booleans?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/dao/PartitionMetadataAdminDao.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/dao/PartitionMetadataAdminDao.java:103]: Could we use a `Dialect` instead of a `boolean`?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/dao/PartitionMetadataDao.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/dao/PartitionMetadataDao.java:68]: Could we use a `Dialect` instead of a `boolean`?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/MapperFactory.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/MapperFactory.java:37]: Could we use a `Dialect` instead of a `boolean`?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/ChangeStreamRecordMapper.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/ChangeStreamRecordMapper.java:94]: Could we use a `Dialect` instead of a `boolean`?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/ChangeStreamRecordMapper.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/ChangeStreamRecordMapper.java:329]: All the fields that we are doing `orElseThrow` are mandatory right (just double checking)?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/ChangeStreamRecordMapper.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/ChangeStreamRecordMapper.java:213]: Since we are calling this method with a `ChangeStreamResultSet` now, this is no longer a row, but a `resultSet`. We make it a row by either calling `resultSet.getPgJsonb(0)` or `resultSet.getCurrentRowAsStruct()`. Could you change the variable name here?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/ChangeStreamRecordMapper.java
Comment (nancyxu123) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/ChangeStreamRecordMapper.java:329]: Yes, I doublechecked.

See JSON string for data change record:     Jsonb: {"data_change_record":{"column_types":[{"is_primary_key":true,"name":"SingerId","ordinal_position":1,"type":{"code":"INT64"}},{"is_primary_key":false,"name":"FirstName","ordinal_position":2,"type":{"code":"STRING"}},{"is_primary_key":false,"name":"LastName","ordinal_position":3,"type":{"code":"STRING"}},{"is_primary_key":false,"name":"SingerInfo","ordinal_position":4,"type":{"code":"BYTES"}}],"commit_timestamp":"2022-11-29T11:31:36.008927-08:00","is_last_record_in_transaction_in_partition":true,"is_system_transaction":false,"mod_type":"INSERT","mods":[{"keys":{"SingerId":"2"},"new_values":{"FirstName":"First Name 2","LastName":"Last Name 2","SingerInfo":null},"old_values":{}}],"number_of_partitions_in_transaction":1,"number_of_records_in_transaction":1,"record_sequence":"00000000","server_transaction_id":"NTczMzU4MTQ3MzQyNzAyNjA0Ng==","table_name":"Singers_qqkl32uoQ1Rhk4ZDqW4KGy2f1oGyhLH6oiiXq71B5U24ZLbTylFANYz","transaction_tag":"","value_capture_type":"OLD_AND_NEW_VALUES"}}


See JSON string for child partitions record:     Jsonb: {"child_partitions_record":{"child_partitions":[{"parent_partition_tokens":[],"token":"__8BAYEGX9zUgAABgsBFg1NpbmdlcnNTdHJlYW1fODVxa29vZHBkM3ZOQUFvOAABhIEGCRQ28AANgoCDCMNkAAAAAAAAhASSRgwFhWcxM185NTE5OTgzAAH__4X_Be6gldXMFYb_Be6hGsHih4eAwGQBAf__"}],"record_sequence":"00000002","start_timestamp":"2022-11-29T11:31:35.951359-08:00"}}

Comment (thiagotnunes) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/ChangeStreamRecordMapper.java:330]: Thanks

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/ChangeStreamRecordMapper.java
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/ChangeStreamRecordMapper.java:100]: I don't think we will call these concurrently, but do you know if they are thread-safe?

**Review by thiagotnunes — APPROVED** [review]

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/ChangeStreamRecordMapper.java
Comment (nancyxu123) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/ChangeStreamRecordMapper.java:100]: I don't really see any documentation explicitly specifying this class as thread-safe, unfortunately. However, like you said, it doesn't seem like these objects will be accessed concurrently?

Comment (thiagotnunes) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/mapper/ChangeStreamRecordMapper.java:100]: I think not, we create a new mapper in the [Setup](https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/dofn/ReadChangeStreamPartitionDoFn.java#L186-L187) method, which is called per SDF, so we (the implementers) won't call it concurrently, but I don't know if the backend will (Windmill in the case of Dataflow). I would think not.

It would be good to test this under load to verify we don't get any errors

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/NameGenerator.java
Comment (pabloem) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/NameGenerator.java:28]: does this make the IO incompatible for pipeline updates? (maybe it already was?)

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/NameGenerator.java
Comment (pabloem) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/NameGenerator.java:50]: I am guessing this is not necessarily a problem, since we have very few instances of metadata tables being created, but cutting UUID can cause issues sooner (https://stackoverflow.com/questions/4564112/is-it-safe-to-turn-a-uuid-into-a-short-code-only-use-first-8-chars)

**Comment by pabloem** [comment]
lgtm - I can merge once tests pass.

**Comment by pabloem** [comment]
Run Java_Examples_Dataflow_Java17 PreCommit

**Comment by pabloem** [comment]
fyi @kennknowles we're working with @nancyxu123 to have this in for 2.44.0 - we'd love your help to coordinate with us on cutting after this merge or coordinating with us to do a cherry pick

**Comment by pabloem** [comment]
LGTM thanks!

**Pull request merged**
